greedy/homework6.tex

```diff
@@ -0,0 +1,225 @@
+\subsubsection*{Proof by Exchange:}
+Let $Alg$ be the process by which the above algorithm operates.
+Assume there exists an input $I$ such that $Alg(I)$ is 
+incorrect.  Let $Opt(I)$ be the optimal result for $I$
+that agrees with the greatest number of steps with $Alg(I)$.
+$Alg(I)$ and $Opt(I)$ must have a first point of disagreement.
+Label this time $t_i$.\\\\
+At time $i$, $Alg(I)$ must have evicted some page $u$ and 
+$Opt(I)$ has some page $v$ evicted.  Since $Alg(I)$ always
+picks the page that will be used furthest in the future,
+the next time page $u$ will be used must be further into
+the future than the next time page $v$ is used.  Let these
+times be respectively $t_u$ and $t_v$.\\\\
+Since $Alg(I)$ and $Opt(I)$ both agreed upto time $t_i$,
+pages $u$ and $v$ must be in the fast memory of $Opt(I)$
+at $t_i$.  Therefore, define $Opt'(I)$ as $Opt(I)$ except
+at time $t_i$ evict page $u$ instead of $v$.\\\\
+To show that $Opt'(I)$ is at least as optimal as $Opt(I)$     
+it needs to be recognized that since $t_u > t_v$, page $v$
+will have to be brought back into memory at least $1$ more
+time before page $u$ is brought back into memory.\\\\
+Since $Opt'(I)$ is at least as optimal as $Opt(I)$ and 
+agrees with $Alg(I)$ for $1$ more step a contradiction has
+be reached.  Therefore, there is not input for which $Alg$
+is incorrect.   
+\subsection*{Problem 17:}
+A greedy algorithm that will determine if there is enough
+information available to fairly partition the goods is as 
+follows:\\\\
+Given a set of goods $\mathcal{G}={G_1,\ G_2,\dots,G_n}$,
+and two orderings on this set, $\mathcal{H}=G_a > G_b >
+\cdots > G_n$ and $\mathcal{W}=G_i > G_k > \cdots > G_m$,
+partition the goods by initially giving $\mathcal{W}$ and 
+$\mathcal{H}$ their maximal elements.
+Remove the element given to the opposing list from each list.
+If this is not possible
+because the maximal elements for both $\mathcal{W}$ and
+$\mathcal{H}$ are the same, conclude \textit{that there is not
+enough information guarantee a fair partitioning}.\\\\  
+While there are elements in $\mathcal{G}$, continue
+giving $\mathcal{W}$ and $\mathcal{H}$ their second
+most desired element.  Ties are broken arbitrarily.  For 
+example, if at a given step both $\mathcal{W}$ and $\mathcal{H}$
+want $G_c$, arbitrarily give $G_c$ to $\mathcal{W}$ and
+cross it off in the ordering for $\mathcal{H}$ and give
+$\mathcal{H}$ the next element in the ordering.\\\\
+After all elements have been assigned,first evenly parition  
+$\mathcal{H}$ and $\mathcal{W}$ into two equal portions.
+If this is not possible because $|\mathcal{G}|$ is odd,
+conclude that \textit{there is not enough information available
+to fairly partition the goods}.  If the upper half of  
+$\mathcal{H}$ and $\mathcal{W}$ is full, conclude that 
+\textit{there is enough information to fairly partition}.\\\\
+If the upper half of $\mathcal{H}$ or $\mathcal{W}$ are
+not full, begin iterating through both orderings and 
```

Comment:
This was a correction to my statement that you can have oxox but not ooxx, which is obviously false. I was wrong when I was talking about the regular expression for checking fair partitionability.